packages/react-components/react-utilities/src/compose/types.ts

```diff
@@ -169,15 +169,24 @@ export type ExtractSlotProps<S> = Exclude<S, SlotShorthandValue | null | undefin
 export type ComponentProps<Slots extends SlotPropsRecord, Primary extends keyof Slots = 'root'> =
   // Include a prop for each slot (see note below about the Omit)
   Omit<Slots, Primary & 'root'> &
-    // Include all of the props of the primary slot inline in the component's props
-    PropsWithoutRef<ExtractSlotProps<Slots[Primary]>>;
+    // Include all of the props of the primary slot inline in the component's props,
+    // omits the props that would collide with slots, e.g: content
+    // excluding the Primary property itself, in case root is another component, e.g: DatePick
+    OmitWithoutExpanding<PropsWithoutRef<ExtractSlotProps<Slots[Primary]>>, Exclude<keyof Slots, Primary>>;
```

Comment:
> Since this only affects a small number of components, it would be ideal if we could avoid adding complexity to the core ComponentProps type. Also, whenever we add non-configurable/"magic" behavior to these types, they become harder to understand (they're already pretty bad in this respect).
>
> Could we instead create a solution that lets components omit specific intrinsic props when they conflict with slot names?

While I completely agree about the complexity of types, I am uncertain whether the approach involving specific props is a viable solution 😶

- We do not maintain `@types/react`, and these types do not adhere to semver ⬇️
- There are no assurances that new properties will be added/removed

In summary, if we manage specific properties without a guard, we risk encountering issues as seen in #29596.  IMO, this is a significant drawback as it does not ensure a future-proof solution.

Reply:
P.S. If I need to pick an option, I would pick 1 as it adds nothing new.

Reply:
I guess I wasn't clear enough in what I meant, let me rephrase it: All options provided here  (Option 1 to Option 3) will have the same [Distributive Conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) issue, as all of them rely on us externalizing `Omit` to the component development!

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAwiZAHZYcwAKxYVABo4AZQA2EGAF84REuQACBcQFduMVcAD0lPDAC0udhC49UZANw0aMAJ5gsYkljbhTGiVPQBeZjThAuGIpAC4xSRgAHjIAE2AANzIRMgAjVRgYUzIAPgCg4x4NcK9ouMTcmmkbLAAPBnh7R2cQVxMzPgFfOHyAeRBgaLdODX4IQSjRF2GPHlLUHJTCmA1cuAAyHsDeYFwAa0np9s9I4XJl1ZzrGkLUeCnWmY64PwAKMC6So-cOscEASheOTgHFU4nENlu8FqLzgh0exx4+UCaB8ACJ0plTGjkXBtNo4E0nMB0BwpCg4AlkOJgLE4B9xlhYHYUOIYAALCCqADm7LggzI6GQcExWQ4uKJ6NF2Nx+LgWAS-KFHDsuNMLBp+x8TFeCsBPmBtwg4iwADpJNzdQl-tUgniCRceHAAO7QPboAjALiEzk8vmhf64x0wbVReIJbQ5aqRoA) of the Option 1 breaking

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAygDYQwAKxYq2uaABMANHHacAvnCIlyAAQJsArlgB2MZcAD0lPDAC0g8BDXqYqMgG46JxkwCiADxhR9E7r1TTZpMopVzLV0qQ2UYYDZgCKxLGhptBAQaREwsEAgAN1i4GAALLHJKAjI4MF4ZYlJ8woBzYGy1OB4IPlyATzAsMTYcTOA1WrhlNWBTdAGYfTgAClRlXDyUdBq4IWBUXChQAeQYLCFh0dM4LZx9w4AjdpSkAEkNbbVUYFwPFr4ASgA6OAAVPIbFBsDgAd3QD1cAxebw4nlaKwgcEuhWUqAOcFB0SW2H03wI0FByCgQmwBG+t1u-wKp1M6wipmQbDgACIsC51EJ0CMANZqCCgtQs2ZMontFZQVSfODssAcaLoZBwACOWigPLgA1yNLF-06WBYW2AYHgeWQaiEvXQggt0TGaiZVJg+tQYhgSPKWEa8BG9vQvmRWAGQ1l5qEBwpiG0NFl0Hgzq6zS8AHVsRBwmSADxcAB8cAAvEUsCUZRyLegeVh2hACM04AB+WZcUv7ctHPkCpqNgDyIGi2bEFGLZDzAC5mtLx1wbPFY4wE4UAMIkSBmDTvLyZjzodmtrniOEffhYQQksQ8UDE9otzkVqs1g+cdCFigQTgjgspOC9-vbwfEd84AAMiTS8oHaHMZzneN9TgZcTDXGBISeGEjy3OEdzLfcNwRARhHPbYQCvG820rata23AsijfGAP3zL8j1TfJ0xgLNnFcdxD03bcAG0LyI8CAF0c0g2cnAYGDE3g1dzDQyjd1vR94T4PCz1AgTrwU0j7wojCqNfd880LL9pNMcwcL4dCnwIsCIOAuCVzMjRkOhV45IwmyNNEmgF3EEgsFMxDKMLJgvzgOAAJgccPEzMh1kyMhB0ucJ3TUEcwtpDRzHraK4Vi+L0skMSJI6RMWH8wLZK8KiTMcoKuIRLcKrq8yMLzECvx-GBM0q5zHlc3A5OahDWqfHNBxtVtaO8m1UHgcqQAClqNComZygRaLhpkjQj2lfM8zUZQQXiWb4CcKimsW3qYAytB8xZZKYFSlkMu0bRSsKIF+XgJVMiZYBDnWrpYGvJlmOUWolmiMhFWRFKTkyNhkABspeCwEGMoXe7Hue173q9TVFTUG5wvC0xFyiXAeXzJgZi9Pa81miBem+DhajpzJPiK7QcyAA) of the Option 2 breaking

Here's a [playground](https://www.typescriptlang.org/play?ts=4.4.4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFCiSxwDeAwiZAHZYcwAKxYVABo4AZQA2EGCIlTsHACZYoAMQCuHfMAgcAvnCIlyAAQLi13GGuAB6SnhgBaXOx2XUZANw0aNhAhpETCwQCAA3LHQYAAssckoCMjgwAQNiUhi4gHNgCI44fghBOBgATzAsEXEcMOAOLLgNbQ50OpgHOAAKVDVcaJQo2LgFYFRcKFA65BgsBUaOZrhxnBm5gCNSwKQASR4JluBcWT4BVABKADo4ABVo0ZRxSQB3dF2YfdRD48LimAg4NZxNSoWZwJ7AGLBBwXAjQJ7IKAKbAEC5bLY3IYuRQQ5rIcRwABEWAAHjNFOgNABrDgQJ4cAldPHw0pRKAWM5wElgSQQ9DIOAAR2sUEpcDqJSGzJu5SwonGwDA8GiyEU1XQWJGMFx4nRZQqwhK-xSWDy8CaOnQhlIgLqDS5KqUClRiBsNC50HgeriP1QAHUIdEIGoYCw7uIFJQOAAeXgAPjgAF5yH1gOHI0kSWSFOhKVhShACAU4AB+Lq8Tmk7jZ+bU2n5UsAeRAEJjIjIKbT3DI8YAXAUOX3eN5fP50QAJa4AWQAMpzqiBLCUZYNpihKHAafA8c9QX84MqIkswxHuM6AkFkcpuLgsH3ojAYIIezYbEoIpIKlALqEAF6p8TIBc0BZDY3COAAqqIr4QLgqA2AA4pIqCoAipQ2AAouAZQAPpYPOlhbK6XpwJhiqlG8HyHOh+E8OgCaBHAAA+5AIjgtBwEx5BrGgWDsZxZBrFAfHMe2EDiMJ5AhICCgSWQ0RCQxImgFksl1GAwayeIdSUrJC7tLJYAIsgICyagQZQDesnvHgOmKeQTyCV4Ph+OeSBjnhn5LhUBjQMwxhaRworHLoVyThAIwEMAkRwAABgAUqIAAaFwUXUny4NRITuAA2tcMoALoxT26JINgoQROg1RZHgpRwKg7y2nACRnkEAAiozbrS6rHpGPlQJyWG1e0WSVcAubkMpZBnkRMpwKlBwZTRJxFKgUZ5d5mZVjmeYFnACXJfN6WZQutHxkm61xJt5IkYNh1UUtqAMaWPr+jEQYhj13BRtg0IvQGyJRvtKV7Gl91ZbRuUFbGsYMX2P34Bcf0xADQN3Yt4MwKgkMVPlsbDi544ecoXlxLC-VMP52liJIMAhTcyC5oMcQpCtYJxPCPCGnVahgAwhC+fyqA0ygigoAo1aZCgayphCtVk7FHYnhwMW6v8gIoE1VbEwQGhaDoEqru0jMShAILJKc000MRr3RMc8hKKoutajoManGdBSnHAABkzAMYrkbFn2nQ+hWWboEwR6pkrgc1jSdJwPoz2nNl7afRwZD5XAfZcBEUAcsxdta47mjO9GPp4zQujDu6jDEWj3ynGts1XdWub5oWqMgwtx3uGdDEcRdodbSLmwcWPY+lhHaAx4P+g+zbheKMo6gl80UZoz3PA+k3FTQ-349ZxuJrKD4xGiCQWBsOAbg8McdG++PxBSH2xxRk2Lb1zT29kCMYRJCJaxgx-HTrGNsWIyQwG7DDce4DLAvxplGH+uRuyVx8DXT0s1z4LivpwSwIckw4JvstQQUYsGX1cFwW+NNUAwxoFieqYgL6EMofAJMnQWZPkYdgiheDTgcgTPGDgahHg+HofAYkiY4CkKYTwng+80AJgJIAh8OgCT7xfCTMU6BNwazCHiYAcwOHKDKA8N6agsj9AhGQPkAIgE6H3l6RRyjgFqPHhok0WiR77x0CwLSuBKQJiYJ0E0-D4z0LElgC4kgsjBLCGcXQ6ibBLB0BAsE0BKSWjqHEMxFis5nH3rAnggSoy-xsLGBJZSgA) of the Option 3 breaking

The problem here is with externalizing the possibility of modifying unions. When we do it internally we can ensure that the distributive conditional types will be properly applied to ensure unions will not be expanded. Any sort of modification in an union type will have this problem! And since we do have unions on the root type of some of our components, I'm not confident that having this is a solid solution long term.

I must say it again though, although this breaks for unions cases, the components with the conflicting types **DO NOT** have unions in their base type, so this will not be a problem **AT THE MOMENT**

Reply:
As this is quite explosive, and it can break at any time due to the [Distributive conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types) I propose we keep it as it is in the PR implementation right now, let's keep any modification of a possible union internal to our type system, to ensure we're properly treating this edge case.

Reply:
Apologies for the delay, this fell off my radar.

I still think we should go with Option 1, which empowers component types to do the right thing. A component can just as easily use `OmitWithoutExpanding` instead of `Omit` in cases where it matters. And unless I'm misunderstanding, it's a bugfix to replace `Omit` with `OmitWithoutExpanding` if the component author forgot, right? (It's not a breaking change).

Regarding @layershifter's comment about `@types/react` potentially breaking in the future... That may be an issue, but it seems like this PR would only help in a very narrow set of circumstances, and there would probably be other breaking changes as well. I don't think we should design around hypothetical future breaks from `@types/react`.

Reply:
Ok, I'll converted this PR to go with Option 1